x/staking/keeper/msg_server.go

```diff
@@ -247,35 +257,74 @@ func (k msgServer) Delegate(goCtx context.Context, msg *types.MsgDelegate) (*typ
 // BeginRedelegate defines a method for performing a redelegation of coins from a delegator and source validator to a destination validator
 func (k msgServer) BeginRedelegate(goCtx context.Context, msg *types.MsgBeginRedelegate) (*types.MsgBeginRedelegateResponse, error) {
 	ctx := sdk.UnwrapSDKContext(goCtx)
+
 	valSrcAddr, err := sdk.ValAddressFromBech32(msg.ValidatorSrcAddress)
 	if err != nil {
 		return nil, err
 	}
+	valDstAddr, err := sdk.ValAddressFromBech32(msg.ValidatorDstAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	srcValidator, found := k.GetValidator(ctx, valSrcAddr)
+	if !found {
+		return nil, types.ErrNoValidatorFound
+	}
+	dstValidator, found := k.GetValidator(ctx, valDstAddr)
+	if !found {
+		return nil, types.ErrNoValidatorFound
+	}
+
 	delegatorAddress, err := sdk.AccAddressFromBech32(msg.DelegatorAddress)
 	if err != nil {
 		return nil, err
 	}
-	shares, err := k.ValidateUnbondAmount(
-		ctx, delegatorAddress, valSrcAddr, msg.Amount.Amount,
-	)
+
+	delegation, found := k.GetDelegation(ctx, delegatorAddress, valSrcAddr)
+	if !found {
+		return nil, status.Errorf(
+			codes.NotFound,
+			"delegation with delegator %s not found for validator %s",
+			msg.DelegatorAddress, msg.ValidatorSrcAddress,
+		)
+	}
+
+	srcShares, err := k.ValidateUnbondAmount(ctx, delegatorAddress, valSrcAddr, msg.Amount.Amount)
+	if err != nil {
+		return nil, err
+	}
+	dstShares, err := dstValidator.SharesFromTokensTruncated(msg.Amount.Amount)
 	if err != nil {
 		return nil, err
 	}
 
+	// if this is a validator self-bond, the new liquid delegation cannot fall below the self-bond * bond factor
+	if delegation.ValidatorBond {
+		if err := k.SafelyDecreaseValidatorBond(ctx, srcValidator, srcShares); err != nil {
+			return nil, err
+		}
+	}
+
```

Comment:
ditto: SafelyDecreaseValidatorBond updates the srcValidator in store but not the data structure.

Reply:
would it make more sense to pass a pointer instead of a copy?

Reply:
@xlab Actually fixed this issue with a pointer in the [LSM repo ](https://github.com/iqlusioninc/liquidity-staking-module/pull/117#pullrequestreview-1519736098)

I'll pull these changes into this branch